File: problems/0036_valid_sudoku/0036_valid_sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Set<Character>[] rowSet = new HashSet[9];
        Set<Character>[] colSet = new HashSet[9];
        Set<Character>[] boardSet = new HashSet[9];

        // initialize set
        for (int i = 0; i < 9; i++) {
            rowSet[i] = new HashSet<>();
            colSet[i] = new HashSet<>();
            boardSet[i] = new HashSet<>();
        }

        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[r].length; c++) {
                char val = board[r][c];
                if (val == '.')
                    continue;
                int boardIdx = (r / 3 * 3) + (c / 3);
                if (!rowSet[r].add(val))
                    return false;
                if (!colSet[c].add(val))
                    return false;
                if (!boardSet[boardIdx].add(val))
                    return false;
            }
        }
        return true;
    }
}

File: problems/0036_valid_sudoku/0036_valid_sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boards = [set() for _ in range(9)]
        for r in range(9):
            for c in range(9):
                val = board[r][c]
                if val == '.':
                    continue
                board_idx = (r // 3) * 3 + (c // 3)
                if val in rows[r] or val in cols[c] or val in boards[board_idx]:
                    return False
                rows[r].add(val)
                cols[c].add(val)
                boards[board_idx].add(val)
        return True